### ISSUE TITLE
startLoggingWithObserver IndexError

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3289]
|<img alt="gtaylor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| gtaylor reported|
|-|-|
|Trac ID|trac#3289|
|Type|release blocker: regression|
|Created|2008-06-10 12:42:08Z|
|Branch|https://github.com/twisted/twisted/tree/logging-recursion-3289|

I'm getting the following error when trying to start an older server app that used the previous major revision of Twisted. I couldn't find anything in the release notes about a backwards-incompatible change that would break things, but perhaps I missed something:

```
2008-06-10 08:21:31-0400 [-] Log opened.
2008-06-10 08:21:31-0400 [-] --------------------------------------------------
2008-06-10 08:21:31-0400 [-]  Command Aliases Loaded: 9
2008-06-10 08:21:31-0400 [-]  Evennia Test Site started on port(s):
2008-06-10 08:21:31-0400 [-]   * 4000
2008-06-10 08:21:31-0400 [-] --------------------------------------------------
2008-06-10 08:21:31-0400 [-] Loaded.
2008-06-10 08:21:31-0400 [-] Traceback (most recent call last):
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\scripts\twistd.py", line 21, in <module>
2008-06-10 08:21:31-0400 [-]     run()
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\Lib\site-packages\twisted\scripts\twistd.py", line 27, in run
2008-06-10 08:21:31-0400 [-]     app.run(runApp, ServerOptions)
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\Lib\site-packages\twisted\application\app.py", line 614, in run
2008-06-10 08:21:31-0400 [-]     runApp(config)
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\Lib\site-packages\twisted\scripts\twistd.py", line 23, in runApp
2008-06-10 08:21:31-0400 [-]     _SomeApplicationRunner(config).run()
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\Lib\site-packages\twisted\application\app.py", line 335, in run
2008-06-10 08:21:31-0400 [-]     self.startLogging(self.getLogObserver())
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\Lib\site-packages\twisted\application\app.py", line 367, in startLogging
2008-06-10 08:21:31-0400 [-]     log.startLoggingWithObserver(observer)
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\Lib\site-packages\twisted\python\log.py", line 519, in startLoggingWithObserver
2008-06-10 08:21:31-0400 [-]     msg("Log opened.")
2008-06-10 08:21:31-0400 [-]   File "C:\Python25\Lib\site-packages\twisted\python\log.py", line 246, in msg
2008-06-10 08:21:31-0400 [-]     o = self.observers.pop(i)
2008-06-10 08:21:31-0400 [-] IndexError: pop index out of range

```
The root Python file for the project is at:
http://code.google.com/p/evennia/source/browse/trunk/server.py

The startup script is at:
http://code.google.com/p/evennia/source/browse/trunk/startup.bat?r=274

Regression? Bug?

Attachments:

* [minimal.tac](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/be9/be927db22afaa709de1f069225b826c5eb6fc788/d763e56d957a03e5b6b87a717d2ee3eca547e359.tac) (187 bytes) - added by exarkun on 2008-06-10 13:20:33Z - minimal reproduction of the problem
* [broken-logging.tac](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/be9/be927db22afaa709de1f069225b826c5eb6fc788/6186c4e0cff6ae524d2c4ecef81bd901b21d3d86.tac) (413 bytes) - added by glyph on 2009-04-20 16:23:42Z - A new version of minimal.tac which opens a failsafe log to observe the traceback that the main log observer produces.
* [really-minimal.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/be9/be927db22afaa709de1f069225b826c5eb6fc788/4d4f9a85769b6f680f76016b5e0bbb445c84a157.py) (373 bytes) - added by glyph on 2009-04-20 16:47:45Z - A ''really'' minimal example, which doesn't involve twistd, and just shows you the full 'maximum recursion depth exceeded' traceback.

<details><summary>Searchable metadata</summary>

```
trac-id__3289 3289
type__release_blocker__regression release blocker: regression
reporter__gtaylor gtaylor
priority__high high
milestone__ 
branch__branches_logging_recursion_3289 branches/logging-recursion-3289
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1213101728000000 1213101728000000
changetime__1266620727000000 1266620727000000
version__None None
owner__ 
cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

This seems to be due to `server.py` initializing logging itself, including clobbering `sys.stdout` (since it does not pass `False` for the `setStdout` parameter of `log.startLogging`).  With the `-n` command line argument, `twistd` then also tries to initialize logging, including clobbering `sys.stdout`.  The interaction between these two results in infinite recursion in the logging system (although the exact reason for this is tricky to debug, since stdout has been well and truly mangled).  Infinite recursion results in an error in a random place which probably results in the logging system trying to remove what it has decided is a buggy log observer (there's another ticket open for fixing this behavior).

If you run without `-n`, then things seem to work fine.  However, I'd recommend removing the `startLogging` call from `EvenniaService.__init__`.  It's the job of `twistd` to handle logging; application code shouldn't be touching that stuff.  You can specify which logfile to use with the `--logfile` option to `twistd`.  We're also working on a way for `.tac` files (`server.py` is really a `.tac` file - it should probably be renamed to reflect this) to specify the log file name programmatically so, so you could continue to load it from your settings module.

It's not clear to me why this worked in any previous version of Twisted; it was probably accidental.  It may be worth fixing this behavior so that two `startLogging` calls that set stdout again work, but it's a pretty severe corner case.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

Okay, here's how the infinite recursion happens.

The first `startLogging` call creates a new log observer which writes to `sys.stdout`, which is at this point the real stdout object.  It also replaces `sys.stdout` with `StdioOnnaStick` which handles calls to its `write` method by emitting a log message.

The second `startLogging` call creates a new log observer which writes to `sys.stdout`, which at this point is the `StdioOnnaStick` instance.  It also replaces `sys.stdout` with another `StdioOnnaStick` instance.

This second log observer is terminally defective.  It handles log events by emitting log events, resulting in infinite recursion, and the traceback the attached example produces.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

In 8.1, the tac file is executed before `twistd` calls `startLogging`.  It creates an observer which refers to the file it gives and replaces `sys.stdout` with a `StdioOnnaStick` instance.  `twistd` then calls `startLogging` which creates an observer which refers to the `StdioOnnaStick` instance.

In 2.5, `twistd` calls `startLogging` before it executes the tac file.  It creates an observer which refers to the real stdout object and replaces `sys.stdout` with a `StdioOnnaStick` instance.  Then the tac file calls `startLogging`, but it creates an observer which refers to the given file.  The value of `sys.stdout` is irrelevant at this point, because the tac file doesn't use it.

So that's why it's broken in 8.1 and works in 2.5.

**Comment by twisted-trac** [comment]
|<img alt="gtaylor's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|gtaylor commented|
|-|-|

I see that -n is now deprecated (for backward compatibility). Running without it is fine, but does that kill my ability to daemonize? Does it only daemonize now if you give it a logfile other than stdout?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Where did you see that `-n` is deprecated?  Are you sure you didn't confuse it with the help for `-q`?

`-n` is short for `--nodaemon` and prevents daemonization.  If you don't specify it, `twistd` will daemonize (and then you **can't** specify stdout as the logfile).  The logfile you specify won't have any effect on whether `twistd` daemonizes or not (except for the previous parenthetical).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

Log observers which raise an exception are no longer removed, so this problem will manifest differently now, but it's still a problem, I think.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

(In [28238]) Branching to 'logging-recursion-3289'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve set owner to @jml|
|-|-|

So it looks like the branch breaks the way trial interacts with logging? It should probably be fixed before merging the branch. Otherwise the change looks sane, I guess.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_11"></a>|@therve set owner to @jml|
|-|-|

+1!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml set status to `closed`|
|-|-|

(In [28266]) Fix infinite recursion error that occurs when calling `startLogging` twice.

 * Author: jml
 * Reviewer: therve
 * Fixes [#3289](https://github.com/twisted/twisted/issues/3289)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
